hugegraph-llm/llm_api/main.py

```diff
@@ -0,0 +1,170 @@
+# specific language governing permissions and limitations
+# under the License.
+
+
+import gc
+import json
+from contextlib import asynccontextmanager
+from enum import Enum
+from typing import Any, Literal, List, Dict
+
+import torch
+import uvicorn
+from fastapi import FastAPI, HTTPException, status
+from fastapi.middleware.cors import CORSMiddleware
+from pydantic import BaseModel
+from transformers import AutoModelForCausalLM, AutoTokenizer, Qwen2ForCausalLM, \
+    Qwen2Tokenizer, GenerationConfig
+from typing_extensions import TypedDict
+
+
+class Message(TypedDict):
+    role: Literal["system", "user", "assistant", "tool", "function"]
+    content: str
+
+
+class ChatRequest(BaseModel):
+    messages: List[Message]
+
+
+class ChatResponse(BaseModel):
+    content: str
+
+
+class Role(Enum):
+    SYSTEM = "system"
+    USER = "user"
+    ASSISTANT = "assistant"
+    TOOL = "tool"
+    FUNCTION = "function"
+
+
+class QwenChatModel:
+    def __init__(self, model_name_or_path: str, device: str = "cuda",
+                 generation_config: GenerationConfig = None):
+        self.model: Qwen2ForCausalLM = AutoModelForCausalLM.from_pretrained(
+            model_name_or_path,
+            torch_dtype="auto",
+            device_map=device
+        )
+        self.tokenizer: Qwen2Tokenizer = AutoTokenizer.from_pretrained(model_name_or_path)
+        self.device = torch.device(device)
+        self.generation_config = generation_config
+
+    @torch.inference_mode()
+    async def achat(self, messages: List[Message]):
+        text = self.tokenizer.apply_chat_template(
+            messages,
+            tokenize=False,
+            add_generation_prompt=True
+        )
+        model_inputs = self.tokenizer([text], return_tensors="pt").to(self.device)
+        generated_ids = self.model.generate(
+            model_inputs.input_ids,
+            max_new_tokens=512,
+        )
+        generated_ids = [
+            output_ids[len(input_ids):]
+            for input_ids, output_ids in zip(model_inputs.input_ids, generated_ids)
+        ]
+        response = self.tokenizer.batch_decode(generated_ids, skip_special_tokens=True)[0]
+        return response
+
+
+def torch_gc() -> None:
+    r"""
+    Collects GPU memory.
+    """
+    gc.collect()
+    if torch.cuda.is_available():
+        torch.cuda.empty_cache()
+        torch.cuda.ipc_collect()
+
+
+@asynccontextmanager
+async def lifespan(app: "FastAPI"):  # collects GPU memory
+    yield
+    torch_gc()
+
+
+def dictify(data: "BaseModel") -> Dict[str, Any]:
+    try:  # pydantic v2
+        return data.model_dump(exclude_unset=True)
+    except AttributeError:  # pydantic v1
+        return data.dict(exclude_unset=True)
+
+
+def jsonify(data: "BaseModel") -> str:
+    try:  # pydantic v2
+        return json.dumps(data.model_dump(exclude_unset=True), ensure_ascii=False)
+    except AttributeError:  # pydantic v1
+        return data.json(exclude_unset=True, ensure_ascii=False)
```

Comment:
What do these two methods do, they don't seem to be used

Reply:
ok, they have been removed.